### PR TITLE
Modify trace for higher performance

### PR DESCRIPTION
**Pull request by gautierronan**
```python
import torch
import timeit

B, N = 10, 10
x = torch.randn(B, N, N)
%timeit x.diagonal(offset=0, dim1=-1, dim2=-2).sum(-1)
%timeit torch.einsum('...ii', x)
# 7.12 µs ± 226 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
# 18.8 µs ± 384 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

B, N = 10, 1000
x = torch.randn(B, N, N)
%timeit x.diagonal(offset=0, dim1=-1, dim2=-2).sum(-1)
%timeit torch.einsum('...ii', x)
# 70.9 µs ± 119 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
# 85.2 µs ± 1.94 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

### CODE REVIEW

File: dynamiqs/utils/utils.py [hunk 1]
```diff
@@ -39,7 +39,7 @@ def trace(x: Tensor) -> Tensor:
         >>> dq.trace(x)
         tensor(3.)
     """
-    return torch.einsum('...ii', x)
+    return x.diagonal(offset=0, dim1=-1, dim2=-2).sum(-1)
```
Comment (pierreguilmin) [inline dynamiqs/utils/utils.py:42]:
```suggestion
    return x.diagonal(dim1=-1, dim2=-2).sum(-1)
```

**Review by pierreguilmin — APPROVED** [review]
Thanks for the idea @abocquet!

**Review by abocquet — COMMENTED** [review]
Since this is performance related, a small before/after benchmark would be nice ☺️

**Comment by pierreguilmin** [comment]
Running the benchmark on my mac for the record:
```python
import torch
import timeit

B, N = 10, 10
x = torch.randn(B, N, N)
%timeit x.diagonal(offset=0, dim1=-1, dim2=-2).sum(-1)
%timeit torch.einsum('...ii', x)
3.02 µs ± 153 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
6.63 µs ± 109 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

B, N = 10, 1000
x = torch.randn(B, N, N)
%timeit x.diagonal(offset=0, dim1=-1, dim2=-2).sum(-1)
%timeit torch.einsum('...ii', x)
9.62 µs ± 150 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
13.6 µs ± 59.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

@gautierronan you definitely need a new machine 😋

**Pull request merged**